### PR TITLE
feat: saved api_key to keychain for user

### PR DESCRIPTION
**Pull request by lifeizhou-ap**
**Why**
Enable users to store api key in keychain for convenience

**What**

***Rules***
When the api key environment variable does not exist,
- it looks up the api key from keychain.  if it does not exist, then
- the users are prompted to enter the api key and
- the users are prompted whether they want to save the api key in the keychain

***Changes***
- Implemented rules as above
- Extract `create_exchange` into a util file
- Moved SessionNotifier to a separate file

Screenshot 2024-10-01 at 4 08 35 PM
Screenshot 2024-10-01 at 4 09 20 PM
Screenshot 2024-10-01 at 4 09 44 PM

Screenshot 2024-10-04 at 3 17 32 PM

### CODE REVIEW

**Comment by lamchau** [comment]
very cool! how would a user update the `keychain` with the same prompts (typo, rotation, etc)?

just had a thought! what do you think of a `goose config` (and maybe even a `goose check`)? that way we can use it as a catch all to do things 
 * check latest version
 * check profiles
 * check api keys of each providers (both environment variables and keychain)?

**Comment by michaelneale** [comment]
cc @aat

**Comment by michaelneale** [comment]
cc @alecthomas

File: src/goose/cli/session.py [hunk 1]
```diff
@@ -52,24 +47,6 @@ def load_profile(name: Optional[str]) -> Profile:
     return profile
 
 
-class SessionNotifier(Notifier):
```
Comment (lifeizhou-ap) [inline src/goose/cli/session.py:55]: moved to a seperate file

File: src/goose/cli/session.py [hunk 2]
```diff
@@ -102,21 +81,6 @@ def __init__(
 
         self.prompt_session = GoosePromptSession()
 
-    def _create_exchange(self) -> Exchange:
```
Comment (lifeizhou-ap) [inline src/goose/cli/session.py:105]: exacted to a util function

**Comment by lifeizhou-ap** [comment]
> just had a thought! what do you think of a `goose config` (and maybe even a `goose check`)? that way we can use it as a catch all to do things
> 
> * check latest version
> * check profiles
> * check api keys of each providers (both environment variables and keychain)?

Yes, that could be useful.

**Comment by lifeizhou-ap** [comment]
> how would a user update the `keychain` with the same prompts (typo, rotation, etc)?

I think we can enhance this later.  When we load the api key from keychain, the console prints the message about the api_key source. If it is not correct, the user still can update by themselves.

**Review by lily-de — APPROVED** [review]

**Comment by alecthomas** [comment]
This is awesome!

**Pull request merged**